### PR TITLE
ref(crons): Allow TIMEOUT complete checkins to have duration updated

### PR DESCRIPTION
**Pull request by evanpurkhiser**
Currently if an IN_PROGRESS check-in times out there is no way to
determine if a later closing check-in  was sent.

For now we can improve the situation slightly by at least updating the
duration of the timed-out check-in.

This is a partial fix for https://github.com/getsentry/sentry/issues/54865

### CODE REVIEW

**Review by davidenwang — APPROVED** [review]

File: src/sentry/monitors/consumers/monitor_consumer.py [hunk 1]
```diff
@@ -293,21 +308,34 @@ def update_existing_check_in(
...
+    updated_checkin["timeout_at"] = get_new_timeout_at(
+        existing_check_in,
+        updated_status,
+        start_time,
     )
 
-    return
+    # IN_PROGRESS heartbeats bump the date_updated
+    if updated_status == CheckInStatus.IN_PROGRESS:
+        updated_checkin["date_updated"] = start_time
```
Comment (JoshFerge) [inline src/sentry/monitors/consumers/monitor_consumer.py:336]: worth adding a test to cover this case?

Comment (evanpurkhiser) [reply src/sentry/monitors/consumers/monitor_consumer.py:336]: I'll follow up -- we've had this behavior for a while, this change doesn't actually change this behavior, so it was already uncovered.

File: src/sentry/monitors/consumers/monitor_consumer.py [hunk 2]
```diff
@@ -251,7 +251,22 @@ def update_existing_check_in(
...
+    # closing check-in.
+    already_user_complete = existing_check_in.status in CheckInStatus.USER_TERMINAL_VALUES
+
+    # Check-in has reached a synthetic terminal status and the updating
+    # check-in is attempting to re-open the already completed check-in.
+    #
+    # XXX(epurkhiser): This check allows synthetically completed check-ins
+    # (like timeouts) to be updated by a user complete check-in. See the later
+    # logic for how existing TIMEOUT check-ins are handled.
+    invalid_status_transition = (
```
[the remark this answers is not in the document]
Comment (evanpurkhiser) [reply src/sentry/monitors/consumers/monitor_consumer.py:264]: Just if they send OK or ERROR

**Review by wedamija — COMMENTED** [review]
I just want to put my understanding of this into words to make sure I understand the concept:
If a checkin is in a failed state that we detected (timed out or missed?), and we then receive an explicit checkin status update (ok, failed), we want to update the status and duration?

Is that right though? I thought we would just like to update the duration and leave the status as is in this case, since the checkin is already "complete"?

**Comment by evanpurkhiser** [comment]
I've update this to make it a bit more explicit, give it a look now

**Review by wedamija — APPROVED** [review]

**Pull request merged**